source/simulator/core.cc

```diff
@@ -1678,19 +1678,38 @@ namespace aspect
             free_surface->execute ();
 
           assemble_advection_system (AdvectionField::temperature());
-          build_advection_preconditioner(AdvectionField::temperature(),
-                                         T_preconditioner);
-          solve_advection(AdvectionField::temperature());
+          if (system_rhs.block(AdvectionField::temperature().block_index(introspection)).l2_norm() <= std::numeric_limits<double>::min())
```

Comment:
This is not quite correct (there is -0.0 and denormalized values that I catch this way), see http://cpp.sh/2nbma for example.

> Second, rhs.l2_norm() has physical units,

Yes, we should probably do something like ``1e-16*something_sensible_with_sensible_units``, but I am not confident how to do this robustly so we went for the easy option.